### Reduction operation: the sum of the numbers in the range [0, a)

In [6]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.034867048263549805 seconds
33.6 ms ± 523 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



In [4]:
# Crear la lista con 10^6 elementos empleando range
N = 10**6
lista = list(range(N))

# a) Suma con bucle for
import time
t0 = time.time()
suma_for = 0
for n in lista:
    suma_for += n
t1 = time.time()
print("Tiempo empleando for:", t1 - t0, "s")

# b) Suma con sum
t0 = time.time()
suma_sum = sum(lista)
t1 = time.time()
print("Tiempo empleando sum:", t1 - t0, "s")


Tiempo empleando for: 0.06322288513183594 s
Tiempo empleando sum: 0.005888700485229492 s


In [5]:
import numpy as np

# Convertimos la lista a array numpy
array = np.array(lista)

# a) Suma con bucle for
t0 = time.time()
suma_np_for = 0
for n in array:
    suma_np_for += n
t1 = time.time()
print("Tiempo empleando array y for:", t1 - t0, "s")

# b) Suma con numpy.sum
t0 = time.time()
suma_np = np.sum(array)
t1 = time.time()
print("Tiempo empleando array y sum:", t1 - t0, "s")


Tiempo empleando array y for: 0.11228489875793457 s
Tiempo empleando array y sum: 0.000408172607421875 s


### Comparación de tiempos de ejecución

En este laboratorio se ha evaluado el rendimiento de la operación de reducción mediante diferentes enfoques: un bucle clásico en Python, el uso de listas y el uso de arrays de NumPy. Los tiempos obtenidos muestran una gran variabilidad: el bucle original tarda sobre 0.0349 s, mientras que la suma mediante un bucle sobre una lista es incluso más lenta (0.0632 s). En cambio, la función integrada sum() reduce drásticamente el tiempo hasta 0.00589 s, siendo mucho más eficiente.

### Uso de NumPy y efecto de la vectorización

Al convertir la lista inicial en un array de NumPy, se observa una diferencia significativa según la técnica empleada. Iterar con un bucle for sobre un array resulta ser la opción más lenta (0.1123 s), ya que emplea el intérprete de Python y no se aprovechan las optimizaciones internas de NumPy. Por el contrario, utilizar np.sum() proporciona el mejor rendimiento de todos los métodos (0.000408 s), gracias a que NumPy ejecuta operaciones vectorizadas en código compilado y gestiona eficientemente la memoria.

### Conclusión

Los resultados confirman que los bucles explícitos en Python son poco eficientes para operaciones masivas, mientras que las funciones internas optimizadas, tanto sum() como, especialmente, np.sum(), ofrecen mejoras de rendimiento muy significativas.